### DISCUSSION TITLE
netbox ipam python module doesnot work

### DISCUSSION THREAD

**Discussion by lasag-2511** [netbox-community/netbox#13479] [General]
Hi all,

Anyone uses automatic scan of networks and importing ip address to netbox using python? I am trying some python script to scan and add ip addresses to netbox. But the python module netbox.ipam.create_ip_address - doesnot work anymore. Any suggestions would be appreciated.

**Comment by candlerb** [comment]
> the python module netbox.ipam.create_ip_address

And what module is that exactly? It's not part of Netbox, that's for sure:

```
$ cd /opt/netbox
$ find . -name '*create_ip_address*'
$ grep -R create_ip_address .
$ 
```

Is this some code you've written yourself, or some third-party library you're using?  Is it a component of this "python script to scan and add ip addresses to netbox"?  Is it some random script you found on the Internet?  Very likely it's broken, or worked with an old version of Netbox but not a new one.

> doesnot work anymore

http://www.catb.org/~esr/faqs/smart-questions.html#idm368
http://www.catb.org/~esr/faqs/smart-questions.html#beprecise

You need to be more specific: show the code you're running that "does not work", explain what you think it should be doing, and explain what it does instead (including showing the full, complete error message and/or backtrace, if any, copy-pasted verbatim)

Netbox supports many different forms of external integration, so you also need to be clear what you're using (e.g. the REST API directly, the REST API via some front-end library like pynetbox, the GraphQL API, the Django ORM directly, the Django ORM from a Custom Script or Plugin, etc)

Include information about the versions of all key software you are using, including most importantly Netbox itself, and Python.

**Reply by lasag-2511** [reply]
Hi Thanks for the fast response. I am using python-netbox module to create ip addresses. Earlier this was working well and fine. But now netbox.ipam.create_ip_address doesnot work but it justs update ip addresses. I am not sure if there was some issues with python module itself. 

I dont see any errors specifically when running the code. I found the script from github and adjusted according to my environment. I am using the rest api directly and not using any custom scripts or plugins. 

I am using netbox version v3.5.7 and python Python 3.8.10

```
        if netboxip['count'] == 0:
            # Check if in network exists and not exist in netbox
            if ipaddress in found_ip_in_network:
                # Adding in IP netbox
                netbox.ipam.create_ip_address(str(ipaddress))
            else:
                pass        
        else:
            #If not exists in netbox and network
            if ipaddress in found_ip_in_network:
                netbox.ipam.update_ip(str(ipaddress),status="active")```

**Reply by candlerb** [reply]
> I am using python-netbox module to create ip addresses

Never heard of it. Can you give a link?

I know and use [pynetbox](https://github.com/netbox-community/pynetbox), but that doesn't have any "create_ip_address" method AFAIK; you'd use `nb.ipam.ip_addresses.create(...)`

> But now netbox.ipam.create_ip_address doesnot work but it justs update ip addresses

Probably best to raise it with the author of that module then.

> I dont see any errors specifically when running the code

I notice you're not checking any return value from netbox.ipam.create_ip_address, maybe you are supposed to?  And how do you know that create_ip_address is even being called?

Minor point: when quoting code, please put three backticks (` ``` `) on a line of their own, before and after the code. That preserves the indentation and makes it much easier to read.  That is: Github uses Markdown, not BBcode (`[code]...[/code]`).  You can edit your post to sort this out.

**Reply by lasag-2511** [reply]
https://pypi.org/project/python-netbox/ - This is the link for the module overview.
I notice you're not checking any return value from netbox.ipam.create_ip_address, maybe you are supposed to? And how do you know that create_ip_address is even being called? - No this is not yet even called, because netbox.ipam.update_ip(str(ipaddress),status="active") -- this works

**Reply by candlerb** [reply]
I notice you've raised a similar report here: https://github.com/jagter/python-netbox/issues/61

If you just say "your library doesn't work", apart from being disrespectful, it doesn't give any starting point for helping to narrow down the problem. Maybe *your own* code is using the library in an improper way (regardless of whether it used to work with an older version of something).  In any case, you also haven't identified whether it was an older version of Netbox or an older version of this library where it worked the way you expected.

If you want someone to help you, you need to be specific. Show the minimal amount of standalone code needed to reproduce the problem in a concrete fashion. And you are more likely to get a helpful response if you've done your own troubleshooting. Even demonstrating that it works with Netbox 3.4 but not 3.5 (for example) would be very helpful to the person maintaining this code.

If it turns out that the code isn't being maintained, there are other options, and pynetbox *is* being actively maintained.

**Reply by markkuleinio** [reply]
> pynetbox _is_ being actively maintained.

Are you sure about that? There haven't been new commits from maintainers for months, and contributions don't get reviewed or merged. Also, there is a new [netbox-python](https://github.com/netbox-community/netbox-python) project started, but it hasn't got much progress either.

I personally have mentally prepared myself for forking `pynetbox` for the customer I'm mostly working with when/if I need features that are not supported in current 7.0.1. That won't be ideal, so I don't know yet.

**Reply by candlerb** [reply]
@markkuleinio wrote:
> > pynetbox is being actively maintained.
> 
> Are you sure about that?

No! 😛  All I mean is, it hasn't broken for my use case with Netbox 3.5.

I note that pynetbox and the new netbox-python are both under "netbox-community" which gives them some sort of semi-official status, whereas jagter/python-netbox is completely independent.

**Reply by markkuleinio** [reply]
I'm pleased to see that today there has been lots of maintainer activity in `pynetbox` so it's suddenly looking totally different than last week.

**Comment by lasag-2511** [comment]
yes i rasied there as well in the last week but no replies was received. hence i am asking here.I was using the version v3.2.5 before it worked, but now doesnot work. So i upgraded the netbox and tried to use the same code there and realized that the module is broken. you can find the code here if you would like to test 
https://gitlab.com/Sasha0986/netbox/-/blob/master/updateipv2.py



N

**Reply by candlerb** [reply]
That's not a minimum reproducer.  You need to create a standalone program (not an IP scanner) which reproduces the problem, maybe like this:

```python
from netbox import NetBox
netbox = NetBox(host="192.168.0.160", port=80, use_ssl=False, auth_token="869e49848b645d3b903800ab2d7325725ce1fcb7")
res = netbox.ipam.create_ip_address("1.2.3.4/24")
print(res)
```

Show the result, and whether it does (or does not) create an IP address. It would also be helpful to look at the traffic which is hitting gunicorn (the Netbox webserver) - run this on the Netbox server itself, at the same time as running the client program

```
tcpdump -i lo -nn -s0 -A tcp port 8001
```

This will show if the HTTP request that the client library is making is malformed, and it will also show the error message coming back from Netbox (which very likely will explain exactly what the problem is).

**Reply by candlerb** [reply]
I also observe that your code is a mixture of using the python-netbox library, and making REST API calls directly using the requests library:

```python
        request_url = f"{NB_URL}/api/ipam/ip-addresses/?q={ipaddress}/"
        ipaddress1 = requests.get(request_url, headers = HEADERS)
```

I'd suggest it's better to stick with one or the other. That is, if you're happy to do queries using direct REST calls, you might as well create IP addresses that way as well. Then you don't have to worry about the third party library at all.

**Reply by lasag-2511** [reply]
Hi Thanks for your suggestions, But here with the help of scanner i wanted to scan all the ips which are reachable with in network. when i did run the above simple code, it doesnot create ip but just displays as below

```{'count': 0, 'next': None, 'previous': None, 'results': []}```

I can see the traffic is definitely passing through

```
08:54:14.509665 IP 127.0.0.1.8001 > 127.0.0.1.53616: Flags [.], ack 322, win 510, options [nop,nop,TS val 2627696502 ecr 2627696502], length 0
E..4..@.@.|A.........A.p6.%..........(.....
..wv..wv
08:54:14.549116 IP 127.0.0.1.8001 > 127.0.0.1.53616: Flags [P.], seq 1:418, ack 322, win 512, options [nop,nop,TS val 2627696541 ecr 2627696502], length 417
E.....@.@.z..........A.p6.%................
..w...wvHTTP/1.0 200 OK
Server: gunicorn
Date: Thu, 17 Aug 2023 08:54:14 GMT
Connection: close
Content-Type: application/json
Vary: HX-Request, Cookie, origin
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
X-Request-ID: 448435ba-acbb-4c1c-af83-eeef05ef760f
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 52


08:54:14.549137 IP 127.0.0.1.53616 > 127.0.0.1.8001: Flags [.], ack 418, win 509, options [nop,nop,TS val 2627696541 ecr 2627696541], length 0
E..4H.@.@............p.A....6.'A.....(.....
..w...w.
08:54:14.549156 IP 127.0.0.1.8001 > 127.0.0.1.53616: Flags [P.], seq 418:470, ack 322, win 512, options [nop,nop,TS val 2627696541 ecr 2627696541], length 52
E..h..@.@.|..........A.p6.'A.........\.....
..w...w.{"count":0,"next":null,"previous":null,"results":[]}
```

**Reply by candlerb** [reply]
Firstly, please take your scanner out of the equation: write a standalone program which attempts to create a single IP address in Netbox, similar to the one I [outlined before](https://github.com/netbox-community/netbox/discussions/13479#discussioncomment-6740945) (but haven't tested - that's up to you)

Secondly, we're getting somewhere with the tcpdump, but you've missed the most important part, which is the HTTP *request* which was sent by the library.  It's necessary to see whether it was a GET/POST/PATCH, what the path was, the headers, and the body of the request.

For example, if it starts `POST /ipam/ip-addresses/ HTTP/1.0` then we need to see that and everything from that point onwards.

Then it should be clear why the response is `{"count":0,"next":null,"previous":null,"results":[]}`

**Reply by lasag-2511** [reply]
Hi I have tested your simple code with out scanner to add ipaddress. The same output I have posted on the above quotes with tcpdump and code with out scanner. With Tcpdump I dont see anything. Can i use netbox-python modules or pynetbox for scanning and adding ipaddress to netbox?

**Reply by candlerb** [reply]
This is my last reply.

> The same output I have posted on the above quotes with tcpdump and code with out scanner

I asked that you show the first part of the HTTP exchange - you have not.

>  Can i use netbox-python modules or pynetbox for scanning and adding ipaddress to netbox?

Scanning, and adding IP addresses to Netbox, are two completely separate and independent activities.

**Reply by lasag-2511** [reply]
> This is my last reply.

sorry i have missed the first part of tcpdump. 

> I asked that you show the first part of the HTTP exchange - you have not. -- Its a get request 
```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on lo, link-type EN10MB (Ethernet), snapshot length 262144 bytes
09:47:02.355002 IP 127.0.0.1.59938 > 127.0.0.1.8001: Flags [S], seq 248780666, win 65495, options [mss 65495,sackOK,TS val 2717264347 ecr 0,nop,wscale 7], length 0
E..<..@.@..*.........".A...z.........0.........
..).........
09:47:02.355027 IP 127.0.0.1.8001 > 127.0.0.1.59938: Flags [S.], seq 4021659945, ack 248780667, win 65483, options [mss 65495,sackOK,TS val 2717264347 ecr 2717264347,nop,wscale 7], length 0
E..<..@.@.<..........A."...)...{.....0.........
..)...).....
09:47:02.355050 IP 127.0.0.1.59938 > 127.0.0.1.8001: Flags [.], ack 1, win 512, options [nop,nop,TS val 2717264347 ecr 2717264347], length 0
E..4..@.@..1.........".A...{...*.....(.....
..)...).
09:47:02.355141 IP 127.0.0.1.59938 > 127.0.0.1.8001: Flags [P.], seq 1:322, ack 1, win 512, options [nop,nop,TS val 2717264347 ecr 2717264347], length 321
E..u..@.@............".A...{...*.....i.....
..)...).GET /api/ipam/ip-addresses/ HTTP/1.0
X-Forwarded-Host: 192.168.200.24
X-Real-IP: ::ffff:192.168.200.1
X-Forwarded-Proto: https
Host: localhost:8001
Connection: close
User-Agent: python-requests/2.30.0
Accept-Encoding: gzip, deflate
Accept: application/json
Authorization: Token a98b952b558f0436a2b6e14402bc214ea3ff2497
```

**Reply by markkuleinio** [reply]
`netbox.ipam.create_ip_address()` is not sending GET request:

https://github.com/jagter/python-netbox/blob/master/netbox/ipam.py#L54-L62

**Reply by lasag-2511** [reply]
So it should be post even post doesnot work

**Reply by lasag-2511** [reply]
any suggestions would be appreciated. 
{"count":0,"next":null,"previous":null,"results":[]}' - for me it looks like this part is not working at all with API requests
